design/general-progress-monitoring.md

```diff
@@ -0,0 +1,499 @@
+
+## Workflow
+
+When a Backup or Restore Action is executed, any BackupItemAction, RestoreItemAction, or
+VolumeSnapshot plugins will return operation IDs (snapshot IDs or other plugin-specific
+identifiers).  The plugin should be able to provide status on the progress for the snapshot and
+handle cancellation of the operation/upload if the snapshot is deleted.  If the plugin is restarted,
+the operation ID should remain valid.
+
+When all snapshots have been taken and Kubernetes resources have been persisted to the ObjectStorePlugin
+the backup will either have fatal errors or will be at least partially usable.
+
+If the backup/restore has fatal errors it will move to the "Failed" state and finish. If a
+backup/restore fails, the upload or other operation will not be cancelled but it will not be
+monitored either.  For backups in any phase, all snapshots will be deleted when the backup is
+deleted.  Plugins will cancel any data movement or other operations and remove snapshots and other
+associated resources when the VolumeSnapshotter DeleteSnapshot method or DeleteItemAction Execute
+method is called.
+
+Velero will poll the plugins for status on the operations when the backup/restore exits the
+"InProgress" phase and has no fatal errors.
+
+If any operatios are not complete, the backup/restore will move to either WaitingForPluginOperations
+or WaitingForPluginOperationsPartialFailure or Failed.
+
+Post-snapshot and other operations may take a long time and Velero and its plugins may be restarted
+during this time.  Once a backup/restore has moved into the WaitingForPluginOperations or
+WaitingForPluginOperationsPartialFailure phase, another backup/restore may be started.
+
+While in the WaitingForPluginOperations or WaitingForPluginOperationsPartialFailure phase, the
+snapshots and item actions will be periodically polled.  When all of the snapshots and item actions
+have reported success, the backup/restore will move to the Completed or PartiallyFailed phase,
+depending on whether the backup/restore was in the WaitingForPluginOperations or
+WaitingForPluginOperationsPartialFailure phase.
+
+The Backup resources will not be written to object storage until the backup has entered a final phase: 
+Completed, Failed or PartialFailure
+
+## Reconciliation of InProgress backups
+
+InProgress backups will not have a `velero-backup.json` present in the object store.  During
+reconciliation, backups which do not have a `velero-backup.json` object in the object store will be
+ignored.
+
+## Plug-in API changes
+
+### OperationProgress struct
+
+    type OperationProgress struct {
+        completed bool                          // True when the operation has completed, either successfully or with a failure
+        err error                               // Set when the operation has failed
+        itemsCompleted, itemsToComplete int64   // The number of items that have been completed and the items to complete
```

Comment:
I was thinking we can provide a more flexible `progressJSON  string` so one plugin can report the progress in its own way...  
If the progress has to be reflected in some of the fields of the backup CR we can tell the plugin developer to provide the JSON string in a certain format like `{ 'completed_bytes':xxxx, 'total_bytes':xxxx }`.  And for those plugins do not have the capability to report progress it just returns an empty JSON object `{}`

Do you think this would work?

Reply:
Plugin author could provide two ints and a string for units. Velero could do a quick division for percentage calculations. Or yeah just accept any string.
```
1 processed / 5 total (<plugin-author-chosen-units>)
```

Reply:
@reasonerjt So I thnk you're suggesting we replace both the "completed" and the "total" ints with a single string with the assumption that the plugin will just fill in a user-readable string. In other words, instead of `itemsCompleted=1000;itemsToComplete=10000` we'd have `progressSummary="1000 out of 10000 bytes completed". -- and we would *not* do anything in velero to process this string, we'd just attach it in an appropriate location for progress reporting -- is that the suggestion? I'm OK with that, as long as we're OK with Velero not attempting to give a user an idea of percentage completion.

@kaovilai If we're going to collapse the done/total ints into a single string, I'd rather avoid having to parse this on the velero side, since there's no way velero will know for sure what the plugin might want to report. If we need percentage calculations, then we should stick with some variation on the original design (with updated names as appropriate); if we don't need percentages, then we can just replace the two integers with a single string that Velero will never attempt to parse.